[FEAT] TurboModules guide

lindboe opened the pull request:
This PR is a draft of the TurboModules guide, based on the existing work for the Fabric Components guide. The code here is pulled from https://github.com/react-native-community/RNNewArchitectureLibraries/tree/feat/back-turbomodule as much as possible.

## Testing Status

These steps worked for me on Android, but iOS is failing to load the TurboModule. No build errors, and the generated code looks fine as far as I can tell. I tried renaming the module to `Calculator` as in the original example but still no luck.

I have the module code and app I used here: https://github.com/lindboe/TMGuide

## Still need to fix prior to merge (usually marked by TODO)

* Troubleshooting section: I'd like to add the following sections, but I'm not sure what instructions to write because I get stuck on it myself:
  * How to troubleshoot a native module that won't load 
  * Managing the Android build process if you've made a mistake, went to correct it, and ended up with mismatched native/JS specs, which produces an error

## Comparison with Fabric Component guide

There's a lot of overlap of the ground we want to cover between this guide and the TurboModule guide. I think there may even be an argument for extracting some of it into a separate, shared document (primarily around understanding JavaScript concepts).

I think it's worth going through and comparing the two and picking the preferred versions of the shared text.

These are the major themes of what I changed:
* Reworded to avoid "we" as recommended in the style guide. Preferred "you", and imperative-style sentences for instructions.
* Bulleting:
  * Use numbered steps when describing items that happen in order
  * Either use capitalization and punctuation in list items (when they are full sentences) or use neither (when they are sentence fragments)
* Consistently used `tree` layout for describing folder structure, instead of image
* Added titles to all code blocks (helps re-orient the reader in a long document)
* fixed some errors (grammar, mixed use of `rnt` prefix)
* fix for tab values (the Fabric guide version was crashing)

## Other notes

### Codegen vs CodeGen?

I'm seeing both cases around, which is preferred?

### JS best practices

On the Fabric guide PRs, we talked about wanting to explain JS concepts for devs who may not be familiar. What does this mean for whether we want to demonstrate JS best practices? For example, we explain imports/exports, but don't follow the best practice of a module using an index file to properly export/import its API.

### Document structure

The document gets quite long, and interleaved iOS/Android instructions might make it easier to lose your place. I wonder if it might be beneficial to split the instructions by platform? Obviously then we'd still have some repetition in the JS configuration.

### Explaining CodeGen

I followed the Fabric guide's lead on this, but I noticed that we have phrases like "For further information on how the **CodeGen**, have a look [here](/docs/pillars-codegen.md)", fairly late in the document after we've already mentioned CodeGen several times. I think we might want to refactor that structure.

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
````diff
@@ -3,16 +3,700 @@ id: pillars-turbomodules
+<TabItem value="flow">
+
+```typescript title="NativeCalculator.js"
+// @flow
````
unsure if this should use `strict-local`? the RNNewArch example didn't, so I kept that

cortinico replied:
nope is file wiht just `// @flow`

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,700 @@ id: pillars-turbomodules
+  add(a: number, b: number): Promise<number>;
+}
+
+export default (TurboModuleRegistry.get<Spec>(
```
Just a point of curiosity: Does exporting default like this defeat lazy loading? Is lazy loading obtained a different way? Or is that lazy loading feature entirely on the native side? (I assume not or I'd expect `TurboModuleRegistry.get` to return a promise)

lindboe replied:
From prior experience with things like this, my understanding is that `get` would be called as soon as this was imported in another file, so in the example app it's not being lazily loaded. We'd either have to export a function that calls `get` or conditionally `require` this export to have lazy loading, correct?

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+            │   └── FBReactNativeSpec.h
+            ├── RCTThirdPartyFabricComponentsProvider.h
+            ├── RCTThirdPartyFabricComponentsProvider.mm
+            ├── RTNCalculatorSpec
```
Something that was surprising to me: after I generated this the first time, I realized I'd made a mistake in naming the module (RN instead of RTN), so changed the code and ran this again.

Instead of replacing the generated folder contents completely, it added the new `RTNCalculatorSpec` _alongside_ the `RNCalculatorSpec` that had been removed. If I totally wiped the generated files and re-generated them, `RNCalculatorSpec` was gone. Is this expected behavior?

cortinico replied:
> Instead of replacing the generated folder contents completely

IIRC the generate-artifacts script is not responsible of wiping the output folder, but just of reading the input and producing an output in the specified folder.

On Android: the cleanup is done by the Gradle Plugin. On iOS we should have a similar mechanism (cc @cipolleschi)

cipolleschi replied:
The cleanup mechanism on iOS is there on `main` but not in 0.68 nor in 0.69. :( 
At the moment, the folder has to be removed manually, unfortunately.

lindboe replied:
Should the docs include a warning in case anyone runs into this?

cortinico replied:
> @cortinico, we wipe the `build` folder every time we run the CodeGen on Android?

Answering for future reference: yes we do wipe the generated folder every time the Codegen needs to run:

https://github.com/facebook/react-native/blob/8a33b75f55ba429e9e7a897179557aa8bb49ab35/packages/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/tasks/GenerateCodegenSchemaTask.kt#L49-L52

This is to prevent stale files from being around.

> Should the docs include a warning in case anyone runs into this?

IMHO not. I would not mention this as this is how the whole Android/Gradle environment works (things are not re-executed unless needed). If there is a issue, is most likely a bug that needs to be addressed on our end 👍

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
````diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+```objc title="RTNCalculator.h"
+#import <React/RCTBridgeModule.h>
+
+@interface RTNCalculator : NSObject <RCTBridgeModule>
````
The corresponding file in the Fabric guide also includes `NS_ASSUME_NONNULL_BEGIN` in this file, but the RNNewArch example doesn't. Wasn't sure if that was important to include?

lindboe replied:
addressed

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
````diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+}
+```
+
+### iOS: Create the `podspec` file
````
Point of interest: [The style guide recommends putting Android before iOS](https://github.com/facebook/react-native-website/blob/main/STYLEGUIDE.md#android-before-ios), but I don't think that's been consistently followed. Is that still something we'd like to keep?

cortinico replied:
In this case, I would prefer local consistency over global consistency. IMHO let's keep the same ordering we followed in the other pages. We can re-sort them after we merged all of them 👍

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+To create a TurboModule, we need to:
+
+1. Define a set of JavaScript specifications.
+2. Configure the module and inspect the code created by Codegen.
```
```suggestion
2. Configure the module and let the Codegen generate the scaffolding.
```

scaffolding can be replaced with boilerplate or other similar terms.

I would just not mention 'inspect the code' in the list of necessary steps.

lindboe replied:
I like scaffolding, I think that's very clear about the role it plays, and implies you're expected to fill in more code.

I'm guessing the original intent in the "inspect the code" wording (from the Fabric Guide) was to talk about how we're going to read the scaffolding generated so we understand what native code to write, which does seem like a useful part of the process to convey, to me?

lindboe replied:
Also just a note, but a few different suggestions left mention using "the CodeGen". Just to make sure we're clear, I don't think this is grammatically correct, it should either be "let CodeGen generate" or "let the CodeGen CLI tool (or process, or script, etc) generate..."

lindboe replied:
Addressed, but leaving open for discussion if needed

cipolleschi replied:
> I'm guessing the original intent in the "inspect the code" wording (from the Fabric Guide) was to talk about how we're going to read the scaffolding generated so we understand what native code to write, which does seem like a useful part of the process to convey, to me?

That's correct, that was my intent there. The user will need to read the code generated by CodeGen in order to know how to use the generated types and files.

In this bullet list, I think that both the sentences fit. I will use the one that requires less effort from the reader (I think it is Nico's suggestion)

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+
+## 1. Folder Setup
+
+In order to keep the module decoupled from the app, it's a good idea to define the module separately from the app, and then add it as a dependency to your app later.
```
nit: we can add it that this will make it easier to open-source it as a library afterwards if needed

lindboe replied:
addressed

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+
+export default (TurboModuleRegistry.get<Spec>(
+  'RTNCalculator'
+) as Spec | null);
```
Why is this `| null` needed?

lindboe replied:
This is my understanding of the return value of `TurboModuleRegistry.get` can be: either an object that implements `Spec` or null. Is that incorrect?

lindboe replied:
I might not be understanding: this is the TypeScript example, but that looks like Flow syntax?

cipolleschi replied:
I agree with Lizzi here. For TypeScript examples is better to use TypeScript idiomatic syntax.

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+
+Finally, we invoke `TurboModuleRegistry.get`, passing the module's name, which will load the TurboModule if it's available. 
+
+:::caution
+We are writing JavaScript files importing types from libraries, without setting up a proper node module and installing its dependencies. Your IDE will not be able to resolve the import statements and you may see errors and warnings. This is expected and will not cause problems when you add the module to your app.
+:::
```
Nit: should we move the `package.json` above so this warning is not necessary. Then in the `Module Configuration` section we just focus on configuring the `codegenConfig`?

lindboe replied:
> should we move the package.json above so this warning is not necessary.

that doesn't stop VSCode from generating warnings, for me.

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+There are other macros that can be used to export modules and methods. You view the code that specifies them [here](https://github.com/facebook/react-native/blob/main/React/Base/RCTBridgeModule.h).
+:::
+
+TODO: more description?
```
Asking @cipolleschi here for a feedback

cipolleschi replied:
I would add a small paragraph on the `getTurbomodule:` method:
```
Finally, the `getTurboModule:` is used by the New Architecture to create a proper instance of the TurboModule so that the JavaScript side can properly invoke its methods. The function is defined in (and requested by) the `RTNCalculatorSpec.h` file that is generated by CodeGen.
```
Or something similar.

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+
+This is the last piece of Native Code for Android. It defines the Package object that will be used by the app to load the module.
+
+## 5. Adding the Fabric Component To Your App
```
```suggestion
## 5. Adding the TurboModule To Your App
```

lindboe replied:
addressed, thanks for catching that 🤦‍♀️

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+import RTNCalculator from 'rtn-calculator/js/NativeCalculator.js';
+
+const App: () => Node = () => {
+  const [currentResult, setResult] = useState<number | null>(null);
```
```suggestion
  const [result, setResult] = useState<number | null>(null);
```

lindboe replied:
Why this change, just to make it match the `[x, setX]` pattern? not a big deal, just will need to rename the `const` in the `onPress` to avoid conflict

lindboe replied:
updated

cortinico replied:
> just to make it match the [x, setX] pattern?

Essentially yes. That's a nit, but the rest of docs we have that rely on `useState` follow this convention

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+import java.util.HashMap;
+
+public class CalculatorModule extends NativeCalculatorSpec {
+
```
Let's add a:
```java
public static String NAME = "RTNCalculator";
```

and replace all the strings both here and inside `CalculatorPackage` with this static field.

cortinico commented:
Answering inline to your questions on the PR description:

> These steps worked for me on Android, but iOS is failing to load the TurboModule. No build errors, and the generated code looks fine as far as I can tell. I tried renaming the module to Calculator as in the original example but still no luck.

I've looked at your code on `lindboe/TMGuide` and it looks everything fine on my end. Perhaps @cipolleschi can spot what's the problem here. I suspect there might be something going on the app setup. 

> There's an ordering issue with the steps for Codegen with Android here (and I suspect with Fabric Components as well): trying to run Codegen without the Android source files present doesn't work. But we're presenting the Codegen step as how one would know what native code to write, so I'm not sure how to fix that problem.

Discussed offline. 

> Troubleshooting section: I'd like to add the following sections, but I'm not sure what 
> instructions to write because I get stuck on it myself:
> How to troubleshoot a native module that won't load

This would be valuable, but I think it could be helpful to write it from the 'error message' point of view. Like, "How to resolve when you get the error message xyz".

> Managing the Android build process if you've made a mistake, went to correct it, and ended up with mismatched native/JS specs, which produces an error

Same as above.

> Comparison with Fabric Component guide There's a lot of overlap of the ground we want to cover between this guide and the TurboModule guide. I think there may even be an argument for extracting some of it into a separate, shared document (primarily around understanding JavaScript concepts).

Agree here. In general they were split originally as we wanted to offer separate guide for TM users vs Fabric users (as the initial playbook was already extremely long).

> These are the major themes of what I changed

Thanks for doing all of them 🙏

> Codegen vs CodeGen?

IMHO CodeGen should be preferred as it's essentially the name of a product (like Fabric or TurboModule).

> JS best practices. [...] For example, we explain imports/exports, but don't follow the best practice of a module using an index file to properly export/import its API.

In theory I agree here. I'm also not a JS expert so I can't make a call here, but I could say that I would try to stay focused on the React Native side of things. We don't need to re-explain JavaScript essentially (i.e. no need to explain what `export` does, unless is critical for the sake of the guide). I'm fine with having a couple of sentences for the imports and essentially walk the users through the code. 

> The document gets quite long, and interleaved iOS/Android instructions might make it easier to lose your place. I wonder if it might be beneficial to split the instructions by platform? Obviously then we'd still have some repetition in the JS configuration.

Our hope is to try to align the 2 platforms as much as possible. E.g. in 0.70 the configuration of the two platforms will be aligned. Keeping this in mind, I would try to not split it 

> Explaining CodeGen
> I followed the Fabric guide's lead on this, but I noticed that we have phrases like "For further information on how the CodeGen, have a look [here](https://github.com/docs/pillars-codegen.md)", fairly late in the document after we've already mentioned CodeGen several times. I think we might want to refactor that structure.

Yup that page is going to be a deep dive around Codegen. I think we can move the link a bit earlier just so the reader knows where to find further docs if they need to.

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
 
-This section is a step-by-step guide to create a TurboModule from scratch. The list of subsections is roughly:
+- Strongly typed interfaces that are consistent across platforms
+- The ability to create shared C++ code for use across platforms
```
@cortinico can you elaborate for me more what you want to convey by including "C++-only" here? Is there a misunderstanding you're trying to avoid? To my eye, I wouldn't assume the original wording means that I can share code in other languages as well.

(Side note: it would have to be written "C++-only", which looks a bit odd with the "++-", so I would try to reword this once I have a better understanding of intent.

cipolleschi replied:
I think we are saying two different things:
* In the original formulation, we wanted to convey that we can share code between iOS and Android using C++ code.
* In the suggestion by Nico, we are suggesting that we can create TM that can be used on both platforms and that contains only C++ code (this can be used for TM that do not need to access specific platform's features).

Both statements are true. I would specify both or look for a formulation that implies both meanings. What do you think?

lindboe replied:
Tried to rephrase this to the best of my ability to explain both those ideas (thanks for laying them out so clearly @cipolleschi). Let me know what you think!

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+
+To create a TurboModule, we need to:
+
+1. Define a set of JavaScript specifications.
```
One change: if it's no longer a "set", I would say "the JavaScript specification", not plural.

lindboe replied:
addressed

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+react {
+    jsRootDir = file("../js/")
+    libraryName = "RTNCalculator"
+    codegenJavaPackageName = "com.RTNCalculator"
```
Yeah, I just matched the Fabric guide: https://github.com/facebook/react-native-website/blob/fd81aa82a7aceb9ff3b10d7b279a3de1055daf94/docs/the-new-architecture/pillars-fabric-components.md?plain=1#L281

cortinico replied:
Yup let's keep it aligned then :+1:

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+As first step for both iOS and Android, this guide shows how to execute manually the scripts used by **CodeGen** to generate the required code. Further information on **CodeGen** can be found [here](/docs/pillars-codegen.md)
+
+:::caution
+The code generated by the **CodeGen** in this step should not be committed to the versioning system. React Native apps are able to generate the code when the app is built. This allows to avoid any ABI incompatibility and to ensure that a consistent version of the **CodeGen** is used.
```
This was mentioned in the Fabric guide: https://github.com/facebook/react-native-website/blob/fd81aa82a7aceb9ff3b10d7b279a3de1055daf94/docs/the-new-architecture/pillars-fabric-components.md?plain=1#L305. Should that also be changed? I'm not really familiar with what "distributing prebuilts" means in this specific context.

cipolleschi replied:
When a distributor builds a library and distribute it to the public so that users don't have to build it themselves, they are creating a prefab.

We have ABI issues when a prefab uses a version of a library that is used also by the app and the version used by the prefab is different from the one used by the app (and potentially incompatible). For example, if a TM uses ReactNative 0.68 and we create and distribute a pre-built version of it and it is used by an app that runs 0.69, we may have issues.

> Should that also be changed?

Yes, I will update it there as well.

lindboe replied:
Updated to be similar to the Fabric guide, but I made a slight edit:

```
The code generated by **Codegen** in this step should not be committed to the versioning system. React Native apps are able to generate the code when the app is built. This allows an app to ensure that all libraries have code generated for the correct version of React Native.
```

I'm not sure if this matches the intent; but I felt "ensure that a consistent version of the **CodeGen** is used across all the libraries." was a bit ambiguous -- all the libraries in what context? I assumed it meant all of a given app's libraries.

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
````diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+
+```sh title="Running CodeGen for iOS"
+cd MyApp
+yarn add ../RTNCalculator
+cd ..
````
> the extra ..

By this, do you mean in the `yarn add` command? That would be needed, yarn would need to be run from `MyApp`

As for the rest, I copied the way the Fabric guide did it.

cipolleschi replied:
We can remove the `cd ..` line, but in that case we have to update the next command as it follows:
```sh
node node_modules/react-native/scripts/generate-artifacts.js \
  --path . \
  --outputPath ../RTNCalculator/generated/
```
I think that the command is clearer if we run it from the parent folder, using `cd ..`, because it uses less `.` and `..` which adds more cognitive load for the reader.. What do you think?

lindboe replied:
I also think it's a bit clearer from the parent. I'd rather be a bit more verbose for clarity's sake.

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';
 
-This section must have a warning that it works only with the new architecture enabled. It points to the [migration section](../new-architecture-intro).
+If you've worked with React Native, you may be familiar with the concept of Native Modules, which allow JavaScript and platform-native code to communicate over the React Native "bridge", which handles cross-platform serialization via JSON.
```
addressed

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+  defaultConfig {
+    minSdkVersion safeExtGet('minSdkVersion', 21)
+    targetSdkVersion safeExtGet('targetSdkVersion', 31)
+    buildConfigField("boolean", "IS_NEW_ARCHITECTURE_ENABLED", "true")
```
addressed

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+
+Of interest in the `build.gradle` file:
+
+- The `defaultConfig` block, within the `android` block, adds a `buildConfigField` to enable the New Architecture.
```
addressed

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+- Run **CodeGen** to see what it generates.
+- Write your native code, implementing the generated interfaces.
+
+When developing a React Native app that uses a TurboMOdule, it is responsibility of the app to actually generate the code using **CodeGen**. However, when developing a TurboModule as a library, we need to reference the generated code, and it is therefore useful to see what the app will generate.
```
addressed

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,699 @@ id: pillars-turbomodules
+
+This script first adds the `RTNCalculator` module to the app with `yarn add`. Then, it invokes Codegen via the `generate-artifacts.js` script.
+
+The `--path` option specifies the path to the app, while the `--outputPath` option tells the script where to output the generated code.
```
addressed

lindboe commented:
>     How to troubleshoot a native module that won't load

> This would be valuable, but I think it could be helpful to write it from the 'error message' point of view. Like, "How to resolve when you get the error message xyz".

For the TurboModule that won't load (like I was encountering with iOS here),there is no error message, it just returns null, and I'm rather stuck on it so I'm not sure I know what to write as advice! 😄 

For the Android codegen failure, shouldn't be too hard to get that error message again to use.

> IMHO CodeGen should be preferred as it's essentially the name of a product (like Fabric or TurboModule).

Happy to update it all to this, but as I was going through and doing that, I noticed that the actual code pieces that refer to CodeGen do it as "Codegen":

```
./gradlew generateCodegenArtifactsFromSchema
```

```
s.dependency "React-Codegen"
```

It might be more confusing to use "CodeGen" in prose if the code consistently uses "Codegen". (For what it's worth I don't think either looks wrong, it's a stylized name so either is acceptable.)

On the overlap of this guide and Fabric:

- there are several items that are now inconsistent, or giving feedback on over here; it might be best to go through and do another pass over both the guides to get the guides consistent with each other. This would help with the questions around structure, grammar, how much to address JS, etc.

lindboe commented:
I pushed an update with the smaller fixes requested. I have not tested the code changes yet, but they seemed relatively safe.

cipolleschi commented:
Answering inline to your questions on the PR description [cit]:

> These steps worked for me on Android, but iOS is failing to load the TurboModule. No build errors, and the generated code looks fine as far as I can tell. I tried renaming the module to Calculator as in the original example but still no luck.

I will look into it after lunch. Hopefully it is just something misconfigured.

> It might be more confusing to use "CodeGen" in prose if the code consistently uses "Codegen". (For what it's worth I don't think either looks wrong, it's a stylized name so either is acceptable.)

That's a good point... I'd say that we can revert it to `Codegen` everywhere for consistency, but I'll do that after we merge everything so that we just have to run the rename only once (I'll create a task for that)
@cortinico what do you think?

> there are several items that are now inconsistent, or giving feedback on over here; it might be best to go through and do another pass over both the guides to get the guides consistent with each other. This would help with the questions around structure, grammar, how much to address JS, etc.

I will update the Fabric Guide this afternoon or tomorrow, trying to align the two.

cipolleschi commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,698 @@ id: pillars-turbomodules
+- The `TurboModule` type, which defines the base interface for all TurboModules
+- The `TurboModuleRegistry` JavaScript module, which contains functions for loading TurboModules
+
+The second section of the file contains the interface specification for the TurboModule. In this case, the interface defines the `add` function which takes two numbers and returns a promise that resolves to a number. This interface type **must** be named `Spec` for a TurboModule.
```
> This interface type **must** be named `Spec` for a TurboModule.

Not sure about this bit. I'll make some test: we are defining an interface and the `TurboModuleRegistry.get` method is generic over some type... so... we should be able to pick the name we prefer, theoretically...

lindboe replied:
I tried to generate it without this name and it didn't work, I think the error message even told me it needed to be `Spec`.

cipolleschi commented on docs/the-new-architecture/pillars-turbomodule.md:
````diff
@@ -3,16 +3,698 @@ id: pillars-turbomodules
+
+The file will look like this:
+
+```ruby title="rnt-calculator.podspec"
````
```suggestion
```ruby title="rtn-calculator.podspec"
```

cipolleschi commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,698 @@ id: pillars-turbomodules
+
+package = JSON.parse(File.read(File.join(__dir__, "package.json")))
+
+folly_version = '2021.07.22.00'
```
React native 69 still comes with `2021.06.28.00-v2`. Not sure whether we want to use the new version: this guide will not work using `2021.07.22`...

cortinico replied:
Yes my bad :( You're actually right. I forgot that this guide is not based upon the current status of `main`. Apologize for the back and forth

lindboe replied:
Addressed

cipolleschi commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,698 @@ id: pillars-turbomodules
+
+The generated code is stored in the `MyApp/node_modules/rtn-calculator/android/build/generated/source/codegen` folder and it has this structure:
+
+TODO: ordering issue. I can't actually get codegen to work at this point. Only works once I add the native files following. Probably similar issue for Fabric guide. Including generated file structure for later ease of use:
```
Do you have errors here or the CodeGen just does not generate anything?

cipolleschi replied:
We went further in the exploration and you were right. Android requires other files in order to be picked up by the Codegen. Specifically, it needs also:
-  The Android Manifest
- A file that implements the `ReactPackage` interface or that extends the `TurboReactPackage` class

lindboe replied:
From the conversation in Discord, I'm not sure still exactly how much code I would write for that `TurboReactPackage` class before running Codegen; especially since I know I've run into issues on the Android side with creating an incomplete or incorrect interface before and then changing it.

cipolleschi commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,698 @@ id: pillars-turbomodules
+                          false, // needsEagerInit
+                          true, // hasConstants
+                          false, // isCxxModule
+                          false // isTurboModule
```
I think this should be `true`. @cortinico what do you think?

cortinico replied:
Yup correct, this should be fixed

cipolleschi reviewed (commented):
Great guide! Thanks for writing it. I'll now look at the repo and see why it is not working!

cipolleschi commented:
@lindboe The problem with iOS was autolinking. Versions of React Native < 0.69 use a version 7.x of the React Native CLI which requires an `xcodeproj` in order to autolink the pod. This requirement has been dropped in React Native CLI 8.x, which is used by React Native 0.69.

I tried to create `AnApp` at the same level of `MyApp` using version 0.69 and the `pod install` step reports:
```
...
Auto-linking React Native module for target `AnApp`: rtn-calculator
Setting USE_CODEGEN_DISCOVERY=1
[Codegen] Adding script_phases to React-Codegen.
[Codegen] Generating ./build/generated/ios/React-C
...
```
and
```
...
Installing hermes-engine (0.69.0)
Installing libevent (2.1.12)
Installing rtn-calculator (0.0.1)
Generating Pods project
...
```

And the app works. Given that we already released 0.69, and that nothing in the guide has an explicit reference to 0.68, we should be good to go.

lindboe commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,698 @@ id: pillars-turbomodules
+
+In order to keep the module decoupled from the app, it's a good idea to define the module separately from the app, and then add it as a dependency to your app later. This is also what you'll do for writing TurboModules that can be released as open-source libraries later.
+
+Next to your application, create a folder called `RTNCalculator`. (**RTN** stands for "**R**eact**T** **N**ative", and is a standard prefix for React Native modules).
```
I think "standard" is also okay to mean what you're conveying here ("a standard" can mean "a required method of doing thing", but it can also mean "conventional"). But "conventional prefix" is also okay, but it sounds more awkward to me.

lindboe replied:
Addressed, chose "recommended" as I think it conveys the idea better

lindboe commented:
>  Given that we already released 0.69, and that nothing in the guide has an explicit reference to 0.68, we should be good to go.

Should we at least address that this is targeted at a minimum version of 0.69? (I know that the plan is to update to current versions as we go, but thinking about it, people may have a reason to not be on latest RN but still want to make a TurboModule, so having versioned docs does seem helpful).

cortinico commented:
> That's a good point... I'd say that we can revert it to `Codegen` everywhere for consistency, but I'll do that after we merge everything so that we just have to run the rename only once (I'll create a task for that)
> @cortinico what do you think?

I don't have a strong opinion at this point 🤔 I'm fine picking one of the two. We can do a pass afterwards and fix everything. Nit: we also have `case-police` set up in the repo. Perhaps we can configure a custom dictionary and add `CodeGen`, `TurboModules` and all the other cases we want to enforce (ref here: https://github.com/antfu/case-police).

lindboe commented:
Updated with the feedback I could address. There are still some potentially remaining items that I want to make sure I call attention to:

For this PR:
* The TabsConstants file update in the fabric branch was causing crashes for me again. I added a commit to fix so this could run (https://github.com/facebook/react-native-website/pull/3168/commits/cc7ab86dec25a0280ce1cfa243318b5842bfdbda) but
  * If that ends up meaning this needs another rebase: I was expecting you all (@cortinico and @cipolleschi) should be able to rebase this doc as needed, since I allowed edits from maintainers, but if that's not the case let me know?
  * I think we might want to still say a minimum supported version of 0.69 somewhere, if this won't work for 0.68. I know many projects don't regularly upgrade, and I think people largely have a perception that 0.68 is where the New Architecture starts being available, so I think they'd expect this to work.

May want to address now or in future:
* My question about the example provided and lazy loading: https://github.com/facebook/react-native-website/pull/3168#discussion_r906430337
* I left some comments above about troubleshooting items (e.g., how to reset Codegen state on Android if you make a mistake) but I'm not sure if they should be addressed (or what the correct fixes are)
* Should we warn about codegen behavior here: https://github.com/facebook/react-native-website/pull/3168#discussion_r916455603

cipolleschi reviewed (commented):
Hi Lizzi! Thank you so much for this effort. It seems very good to me, I'll go over your points of the last comment:

> The TabsConstants file update in the fabric branch was causing crashes for me again. I added a commit to fix so this could run (https://github.com/facebook/react-native-website/commit/cc7ab86dec25a0280ce1cfa243318b5842bfdbda) but
If that ends up meaning this needs another rebase: I was expecting you all (@cortinico and @cipolleschi) should be able to rebase this doc as needed, since I allowed edits from maintainers, but if that's not the case let me know?

I already rebased and fixed the branch. Thanks for your patience here.

> I think we might want to still say a minimum supported version of 0.69 somewhere, if this won't work for 0.68. I know many projects don't regularly upgrade, and I think people largely have a perception that 0.68 is where the New Architecture starts being available, so I think they'd expect this to work.

That's a good point, we could put up some banners for that.

> I was not sure how to address the Android Codegen ordering issue; it sounded like in Discord there was still some question of intended ordering of steps, and it's not 100% clear to me if there's a minimal set of code I can write to get the generation started.

I'll take over on this, don't worry!

> My question about the example provided and lazy loading: https://github.com/facebook/react-native-website/pull/3168#discussion_r906430337

@cortinico do yo know anything more on this?

> I left some comments above about troubleshooting items (e.g., how to reset Codegen state on Android if you make a mistake) but I'm not sure if they should be addressed (or what the correct fixes are)
Should we warn about codegen behavior here: https://github.com/facebook/react-native-website/pull/3168#discussion_r916455603

I'll try to update this as well!

cortinico commented on docs/the-new-architecture/pillars-turbomodule.md:
```diff
@@ -3,16 +3,764 @@ id: pillars-turbomodules
+cd MyApp
+yarn add ../RTNCalculator
+cd android
+./gradlew generateCodegenArtifactsFromSchema --rerun-tasks
```
@cipolleschi Similarly for the Fabric guide, I believe we should not suggest to `--rerun-tasks`.

cortinico reviewed (approved):
Did a final pass and this is great on my end 👍 I'll leave this to @cipolleschi to merge it whenever you wish. Thanks again for helping us with this @lindboe 🙏
